students/kegan/session10/MailroomTools.py

```diff
@@ -0,0 +1,489 @@
+
+    def __contains__(self, donation):
+        """ Returns whether given donation amount
+        has been donated by this donor.
+        Args:
+            donation (int or float) : donation to search for
+        Returns:
+            bool :
+                True if donation is in donation list
+                False otherwise
+        """
+        return donation in self.donations
+
+    def multiply(self, factor, min_donation=None, max_donation=None):
+        donations = map(lambda d: self.apply_factor(
+            d, factor, min_donation, max_donation), self.donations)
+        return Donor(self.name, *donations)
+
+    @staticmethod
+    def apply_factor(d, factor, min_donation, max_donation):
+        within_min = True if min_donation is None else d >= min_donation
+        within_max = True if max_donation is None else d <= max_donation
+        return d * factor if within_min and within_max else d
+
+
+@functools.total_ordering
+class DonorList:
+
+    def __init__(self, *donors):
+        """ Initializes list of donors.
+        Args:
+            donors (args) : donors as arguments
+        """
+        self._donors = [donor for donor in donors]
+
+    @classmethod
+    def from_dictionary(cls, dict):
+        """ Returns DonorList from given dictionary.
+        Args:
+            dict (dic str: list) :
+                dictionary of donor names mapped to donations as list
+        Returns:
+            DonorList : dictionary as DonorList object
+        """
+        donors = [Donor(name, *donations) for name, donations in dict.items()]
+        return cls(*donors)
+
+    @classmethod
+    def read_from(cls, filein):
+        """ Returns DonorList from given TextIOWrapper object.
+        Args:
+            filein (TextIOWrapper) : open file
+        Returns:
+            DonorList : file contents as DonorList object
+        """
+        donors = []
+        for line in filein.readlines():
+            line = line.split(',')
+            try:
+                name = line[0].strip()
+            except IndexError:
+                continue
+            else:
+                str_donations = [
+                    item.strip().strip('$') for item in line[1:]]
+            donations = []
+            for d in str_donations:
+                try:
+                    d = float(d)
+                except TypeError:
+                    continue
+                donations.append(d)
+            donors.append(Donor(name, *donations))
+        return cls(*donors)
+
+    @property
+    def donor_names(self):
+        """ Returns donor names as list.
+        Returns:
+            list of str : list of donor names
+        """
+        return [donor.name for donor in self.donors]
+
+    @property
+    def donors(self):
+        """ Returns donors.
+        Returns:
+            list : list of donors
+        """
+        return self._donors
+
+    @donors.setter
+    def donors(self, donors):
```

Comment:
do you ever have a reason to set the entire donor list?

Reply:
if so, some validation would probably be good.
and if you have a property that simply gets and sets a value -- you don't need a property.